File: crates/core/src/ctx.rs
//! WHAT A MODULE'S LOGIC IS HANDED. `dispatch.rs` owns the routing and
//! CONSTRUCTS one of these per invocation; this file is only its shape.
//!
//! The line between them is the one ADR-006 draws: this file is the SHAPE of a
//! module's capability context — what may be granted, and what every built-in
//! may read — and `dispatch.rs` is the one place that decides, from a
//! manifest's grants, which of it is `Some`. A field's doc comment is where the
//! reason it exists lives; the dispatcher only fills them in.

use kernel::{EventKind, Request, Response, Timestamp};

/// A KV view scoped to one prefix (ADR-006: the handle physically cannot
/// form a key outside its slice). No G4 module is granted Kv yet; the
/// projection-read/write-behind design lands with the first one.
pub struct KvHandle {
    prefix: String,
}

impl KvHandle {
    /// Read one key under the module's prefix (the prefix is prepended here,
    /// not by the caller — callers never spell absolute keys).
    pub fn get(&self, key: &str) -> Option<String> {
        let _ = key;
        todo!("G5: first Kv-granted module")
    }

    /// Stage a write under the prefix; it leaves as an Effect.
    pub fn put(&mut self, key: &str, value: &str) {
        let _ = (key, value);
        todo!("G5: first Kv-granted module")
    }
}

/// The capability context a module's logic receives (§6 `ctx`). Ungranted =
/// `None` = absent, not present-but-refused (ADR-006). Constructed per
/// invocation from the module's effective grants; never stored.
pub struct Ctx {
    /// The handler asked for this agent's conversation to be WIPED. A flag and
    /// not an action, for `emit`'s reason: the window lives on `App` and a
    /// handler holds a projection of it, so the route records the intent and
    /// the dispatcher carries it out (`clear::wipe`).
    pub wipe: bool,
    pub kv: Option<KvHandle>,
    /// Injected time, if granted (I7: even built-ins never read a real clock).
    pub clock: Option<Timestamp>,
    /// Emit events, if granted. PROVISIONAL (G4 discovery): a buffer the
    /// dispatcher drains, not the frozen closure — a closure borrowing App
    /// while the handler also runs against App cannot satisfy the borrow
    /// checker without interior mutability this is simpler than. Also
    /// widened from Custom-only to `EventKind`: the chat module's whole job
    /// is emitting `UserMessage`, which is not a Custom fact.
    pub emit: Option<Vec<EventKind>>,
    /// PROVISIONAL (G4): read-only projections handed to every built-in — the
    /// event kinds so far (views are projections of the log, I8). Becomes a
    /// real capability/section story at G5.
    pub recent: Vec<EventKind>,
    /// Each `recent` entry's injected timestamp (I7), by the same index — a log
    /// a person reads needs the time the fact happened, and `EventKind` has none.
    pub at: Vec<i64>,
    /// Commands the runtime is awaiting right now (`App::running`, R2-8).
    pub running: Vec<String>,
    /// Workspace calls handed to the port and not yet answered (`App::calling`,
    /// R11-4). Oldest first: the workspace runs one at a time, so the head is
    /// the one everything else is queued behind.
    pub calling: Vec<crate::trace::inflight::Inflight>,
    /// What a Stop could do to the call in flight, asked of the engine this
    /// build was composed with (R11-1). A fact like `durable` and stated for
    /// the same reason: one engine really kills a command and the other can
    /// only stop waiting for it, and one button label cannot mean both.
    pub interrupt: kernel::Interrupt,
    /// Agents with an utterance ACCEPTED but not yet pumped — `roster::accepted`'s
    /// window, widened to every agent. It lives in memory only, so a replayed
    /// log has none of it: that is exactly what tells a turn something is
    /// driving from a turn a reload abandoned (12 walk, `chat::fold::driven`).
    pub queued: Vec<String>,
    /// The loaded agent specs (increment 03) — a projection like `recent`.
    pub agents: Vec<agent::AgentSpec>,
    /// The agent files that would not parse — shown, never swallowed.
    pub agent_problems: Vec<String>,
    /// WHAT EACH `model:` KEY REALLY CALLS TODAY: `(the key an agent file asks
    /// for, the catalogue entry that answers it, the model id it sends)`, asked
    /// of `ModelPort::resolves`. Keyed by the KEY and not by the agent, because
    /// that is what the resolution depends on — six agents saying `local` are
    /// one question, and this is rebuilt on every request.
    ///
    /// Empty when the port cannot say (every host test, by default). The card
    /// then prints the file's own words and no model id, which is the one thing
    /// it must not invent.
    pub resolved_models: Vec<(String, String, String)>,
    /// Which of `agents` were AUTHORED IN THIS BROWSER rather than shipped in
    /// `public/agents/`, and BY WHOM — empty for the person at the keyboard,
    /// otherwise the agent that called `write_agent` (increment 11, 11b). A
    /// model can write an agent that runs with real capabilities, so who wrote
    /// one is a fact the page states rather than a difference to infer.
    pub authored: Vec<(String, String)>,
    /// What every agent is doing (increment 06) — a projection like `recent`,
    /// cloned so a handler cannot move a status by writing to it.
    pub board: Vec<agent::AgentRow>,
    /// WHICH agent this process is (increment 07). A conversation-shaped fact
    /// with no name on it belongs to this one; `/chat` projects one agent's
    /// history and defaults to it.
    pub me: String,
    /// What THIS process's agent actually holds — its window (increment 08).
    /// A projection like `board`: the pane can say how much of the
    /// conversation the model still sees, which after a compaction is not the
    /// same thing as how much of it is on screen.
    pub window: Vec<String>,
    /// The shared space this process's agent works in, as last read from the
    /// store (increment 09) — a projection like `board`, so the inspector
    /// cannot show facts the agent's own prompt does not have.
    pub space: Option<agent::Space>,
    /// Whether what the agent writes in its workspace folder is still there
    /// after a reload — `WorkspacePort::durable`, asked of the engine this
    /// build was composed with (increment 18). Two projections state it in
    /// prose, and the sentence has to follow the engine: container2wasm's
    /// filesystem is memory, and a product that keeps promising IndexedDB
    /// after the owner switches engines is lying in the one place a person
    /// checks before trusting it with a file.
    pub durable: bool,
    /// How many of `recent` were replayed from storage rather than appended by
    /// this page. Everything below it happened on an earlier load, in a Linux
    /// that was rebuilt since — and possibly on the OTHER engine, which is why
    /// the scrollback marks those rows instead of showing their output as
    /// current (R10-5).
    pub booted: usize,
    /// Whether this CONTEXT owns the log it is projecting (`log::writership`).
    /// A projection like `board`: the conversation has to say why it will not
    /// take a turn, and the reason is a fact in the log, not a flag the adapter
    /// reached in and set.
    pub writership: crate::log::writership::Writership,
}

/// A tier-0 built-in's logic. A plain fn pointer, not a trait object: no
/// state may hide in a built-in (state lives in the log/store like everyone
/// else's — I9), and fn pointers keep the dispatch table one flat array.
pub type BuiltinHandler = fn(&Request, &mut Ctx) -> Response;

File: packages/core/src/ctx.js
/**
 * WHAT A HANDLER IS HANDED: its effective grant, and exactly what that grant
 * buys. Ungranted is ABSENT — `null`, not present-but-refused — so a handler
 * cannot reach a capability it was denied even by mistake (I6).
 *
 * Rust handed over twenty pre-computed projections here (the board, the
 * roster, the window, every resolved model) because a handler could not borrow
 * `App` while `App` was borrowed mutably to run it. That is a borrow-checker
 * shape, not a problem shape: it made every request clone the whole log, which
 * made each poll dearer than the last. The projections that survive are the
 * ones a view actually asks for.
 *
 * NO HANDLER RECEIVES THE EVENT ARRAY (RULINGS, attack 2). Handing over
 * `log.events` would have been a second authority on history — `push` and
 * `splice` are writes the append-only log itself refuses to offer, and a
 * `readonly` JSDoc annotation cannot stop either. History arrives as `project`,
 * a read of a registered reducer's folded state: the fold ran once, at append,
 * and there is no array here to walk even if a handler wanted to (I20).
 * @module
 */

import { effectiveGrant, grants } from '@harness/kernel'

import { driving } from './drive.js'

/** @typedef {import('@harness/kernel').CapabilityGrant} CapabilityGrant */
/** @typedef {import('@harness/kernel').Fact} Fact */
/** @typedef {import('@harness/kernel').Manifest} Manifest */
/** @typedef {import('@harness/kernel').Timestamp} Timestamp */
/** @typedef {import('./app.js').App} App */

/**
 * @typedef {{
 *   grant: CapabilityGrant,
 *   clock: Timestamp|null,
 *   emit: ((fact: Fact) => void)|null,
 *   project: (name: string) => unknown,
 *   me: string,
 *   driving: (agent: string) => boolean,
 *   durable: boolean,
 *   bootedAt: number,
 * }} Ctx
 */

/**
 * RECORD A FACT AND HAND IT TO THE LOOP. Both, together: a fact the log holds
 * that the loop never sees is a message a person watches land and an agent
 * never answers — the seam is synchronous by construction, so the QUEUE is how
 * a press becomes work the driver picks up after `handle` has returned.
 * @param {App} app @param {Fact} fact
 */
function queue(app, fact) {
  const event = app.log.append(fact, app.ports.clock.now())
  app.pending.push({ at: event.at, turnId: null, fact })
}

/**
 * Build the context for ONE invocation. Never stored: a grant is a fact about
 * this request, and a handler that kept one would outlive the narrowing.
 * @param {App} app
 * @param {Manifest} manifest
 * @returns {Ctx}
 */
export function contextFor(app, manifest) {
  const grant = effectiveGrant(manifest.id, manifest.capabilities, app.available)
  return {
    grant,
    clock: grants(grant, 'clock') ? app.ports.clock.now() : null,
    // The module never sees the timestamp it is stamped with, granted `clock`
    // or not: the LOG stamps a fact, because a fact whose time its author
    // chose is a fact a person cannot trust.
    emit: grants(grant, 'emit') ? (fact) => queue(app, fact) : null,
    // Ungated, unlike `emit`: reading a projection is what answering a GET IS,
    // and a capability that every module must hold to do its job is a line of
    // configuration rather than a boundary.
    project: (name) => app.log.read(name),
    // WHOSE LOOP THIS PROCESS RUNS. A handler that guessed it would answer
    // `/chat` with no agent header for whichever agent it liked.
    me: app.me,
    // Live state a RELOAD DOES NOT HAVE, which is exactly why it is here: the
    // shape of the log survives a refresh and the fetch behind it does not, so
    // "a turn is open" and "a turn is running" are two different questions.
    driving: (agent) => driving(app, agent),
    // Whether anything written to the workspace survives a refresh, asked of
    // the port rather than assumed — `durable()` returning `true` on behalf of
    // an adapter that returned `false` is the defect this build refuses.
    durable: app.ports.workspace.durable(),
    bootedAt: app.bootedAt,
  }
}
